### ISSUE TITLE
SHEX_demo.ino prints garbage (PASS THROUGH MODE)

### ISSUE THREAD

**Issue by brewmanz** [RobTillaart/SHEX#5]
Running SHEX.demo.ino from Arduino IDE Library 
 Version (from SHEX.cpp) "`//  0.2.1   2021-12-28  update library.json, readme, license, minor edits`"
 Prints garbage (raw bytes rather than hex) because it's in pass through mode. 

I'd do a pull request, but not sure of your preferred fix, which could be either:
A) [my preference - why use SHEX if you want pass through?]
  SHEX.h: 
    replace "`  bool      _hexOutput = false;`"
    with  "`  bool      _hexOutput = true;`"
 SHEX.cpp: SHEX::reset():
    replace "`  _hexOutput = false;`"
    with "`  _hexOutput = true;`"
B) 
  SHEX_demo.ino: 
    after "`  SHEX shex(&Serial, 16);`"
    insert "`  shex.setHEX(true);`"

Even with these mods, it doesn't do quite what I want (print "`+0000 41 02 63 64 [A.cd]`"); never mind.

**Comment by RobTillaart** [comment]
Thanks, will dive into this asap.

I labeled it not as bug as I need to think back about the rationale of my choices.

**Comment by RobTillaart** [comment]
> it doesn't do quite what I want (print "+0000 41 02 63 64 [A.cd]"); never mind.

I interpret your remarks as the following requirements:
- default HEX output true 
- add a "printable" column option
- add an address length 4-8 digits option  (think 4 be a better default than 8) ?
  - should there be a ```#define SHEX_COUNTER_DIGITS  4``` so it can be set from command line?

Questions:
- What does the + before the byte count mean?
- do you want square brackets [ ]  around the printable column ?

**Comment by RobTillaart** [comment]
Created a **test** branch to investigate adding a printable column - https://github.com/RobTillaart/SHEX/tree/develop
- added **setCountDigits(digits)** to set the number of digits for the counter in front  digits ∈ { 0, 4, 6, 8 }
  - will definitely be added in a next version.
- derived class SHEXA(), as it has extra memory for the printable buffer.
- works reasonable well, except for the last incomplete line.  (demo sketch)
  - might need a **flush()** function.
- missing is a flag to switch printable column on/off  

```
000000	65 67 1B 6C  4E 85 94 1F  55 0D 0F 40  71 24 7D 77  eg.lN...U..@q$}w
000010	0F 5E 81 01  11 5F 69 68  33 30 8A 17  05 52 66 74  .^..._ih30...Rft
000020	10 57 26 90  33 93 47 80  89 08 7F 93  90 36 3B 83  .W&.3.G......6;.
000030	5F 4B 87 62  8E 31 44 3E  3C 07 90 3A  91 12 71 82  _K.b.1D><..:..q.

...

000100	45 57 53 29  46 70 85 2F  51 26 27 7B  1B 04 6C 00  EWS)Fp./Q&'{..l.
000110	04 6E 73 93  50 82 69 60  07 26 00 4C  55 39 71 0E  .ns.P.i`.&.LU9q.
000120	8F 6E 36 76  39 87 81 73  71 66 43 2B  
```

**Comment by RobTillaart** [comment]
Had an intermediate release 0.2.3 with "VTAB" which was on my list.

```
SHEX VTAB 0

0000	35 7C 82 74  00 90 14 87  10 62 69 21  6B 1A 80 42  
0010	6B 3D 80 72  45 57 53 29  46 70 85 2F  51 26 27 7B  
0020	1B 04 6C 00  04 6E 73 93  50 82 69 60  07 26 00 4C  
0030	55 39 71 0E  8F 6E 36 76  39 87 81 73  71 66 43 2B  
0040	13 75 93 77  2D 03 7B 03  30 55 08 6D  06 50 4A 84  

SHEX VTAB 1

0000	93 0F 5D 4F  81 40 4F 33  7D 59 3A 53  00 27 14 43  

0010	92 26 8F 26  56 79 77 2E  2F 58 6D 57  87 87 3E 86  

0020	27 86 54 75  13 41 5F 7E  6F 19 77 83  03 94 94 59  

0030	68 03 1D 73  29 6B 4A 73  47 04 06 32  67 20 84 5B  

0040	04 4D 43 64  59 73 61 48  4F 75 42 62  70 1D 27 08  

SHEX VTAB 2

0000	80 86 2A 73  6E 11 5A 6B  74 1C 70 0D  77 5A 49 80  
0010	3F 6F 6E 5A  80 00 90 39  06 65 56 1F  5E 27 7E 54  

0020	66 3A 0E 04  13 1C 47 46  71 2F 18 8F  68 3A 83 02  
0030	8A 3F 09 77  1D 38 0B 25  35 95 58 5E  1E 42 85 7C  

0040	56 6D 75 22  59 22 33 95  10 11 0A 59  7F 49 83 5F  
0050	88 7C 87 1F  7A 1C 94 14  57 93 8B 60  8C 3D 70 71  

SHEX VTAB 3

0000	7F 18 22 21  32 7F 95 81  6D 2C 72 33  3E 1D 86 39  
0010	53 64 09 6E  03 5B 79 7D  03 1D 92 03  3D 3A 5B 3F  
0020	14 12 25 4C  16 77 5B 68  33 5C 5F 78  18 0C 1A 75  

0030	4C 0D 5E 79  81 55 4F 58  62 80 75 7B  58 08 81 7E  
0040	44 38 95 69  1D 7E 26 92  45 40 89 74  76 8D 85 94  
0050	14 94 3D 65  44 1B 2E 3D  1C 1C 3F 55  38 08 76 4F  

0060	77 49 1B 2A  35 1C 67 91  15 4B 68 70  20 13 83 23  

SHEX VTAB 4

0000	50 51 55 50  7D 88 2C 91  34 21 1E 38  86 75 83 51  
0010	3A 37 5C 6E  65 01 8E 90  87 37 48 5D  7A 0D 8D 83  
0020	81 44 6E 1D  37 52 0E 84  3F 2D 1C 36  69 3D 1F 3C  
0030	04 65 74 6F  0E 37 16 0D  8F 88 76 2C  8A 2B 27 66  

0040	5D 14 13 3E  75 84 84 61  3B 49 28 45  3C 0B 41 1C  
0050	8E 2D 14 0C  3A 53 63 6D  0D 64 33 0F  34 75 67 31  
0060	74 86 77 2A  35 73 1D 07  52 72 66 04  4C 1B 69 3B  
0070	02 88 16 4F  88 95 4F 62  28 35 56 2B  4D 23 94 25  

0080	11 83 4C 0B  71 40 21 20  48 0B 29 79  8E 85 67 82  
```

**Comment by RobTillaart** [comment]
Updated the develop branch
- added a **flushASCII()** for the **SHEXA** class to dump the ascii column in case the HEX output is reset.
  - not ideal but workable for now.
- replaced the **get/setCounterFlag()** with **setCounterDigits()** which allows  0, 4, 6, 8 digits for the counter.
  - 5 and 7 will be rounded up;  
  - default is 4 => allows up to 65536 bytes which is often enough
- several #defines overrule defaults from command line.

**SHEXA** output could look like   (6 digits for counter, note the space in the ASCII column)
```
000000	65 67 1B 6C  4E 85 94 1F  55 0D 0F 40  71 24 7D 77  e  g.lN...U  ..@q$}w
000010	0F 5E 81 01  11 5F 69 68  33 30 8A 17  05 52 66 74  .  ^..._ih3  0...Rft
000020	10 57 26 90  33 93 47 80  89 08 7F 93  90 36 3B 83  .  W&.3.G..  ....6;.
000030	5F 4B 87 62  8E 31 44 3E  3C 07 90 3A  91 12 71 82  _  K.b.1D><  ..:..q.
000040	6A 0C 8E 73  52 34 43 15  2D 70 15 65  8C 51 26 39  j  ..sR4C.-  p.e.Q&9
000050	10 8C 8C 4F  23 6A 7A 94  2D 45 36 17  8B 6E 37 7E  .  ..O#jz.-  E6..n7~
000060	7B 38 71 14  8A 5E 78 2C  10 22 1A 90  0D 8A 5E 28  {  8q..^x,.  "....^(
000070	32 09 49 93  55 43 7A 27  2F 87 2E 55  49 46 2C 44  2  .I.UCz'/  ..UIF,D

000080	55 0F 4B 86  8E 0B 81 66  90 91 44 92  2A 0F 5B 53  U  .K....f.  .D.*.[S
000090	77 61 03 61  7D 6E 3E 6F  6C 7F 3D 4B  55 44 5F 7E  w  a.a}n>ol  .=KUD_~
0000A0	75 8B 7C 51  06 01 16 6E  2C 22 05 27  07 0F 2B 77  u  .|Q...n,  ".'..+w
0000B0	50 37 69 06  3F 34 1A 3A  63 51 90 26  6C 2F 33 63  P  7i.?4.:c  Q.&l/3c
```

I notice a BUG in the last ASCII column.

**Comment by RobTillaart** [comment]
BUG solved (I guess ;)

```
000000	65 67 1B 6C  4E 85 94 1F  55 0D 0F 40  71 24 7D 77  eg.lN...  U..@q$}w  
000010	0F 5E 81 01  11 5F 69 68  33 30 8A 17  05 52 66 74  .^..._ih  30...Rft  
000020	10 57 26 90  33 93 47 80  89 08 7F 93  90 36 3B 83  .W&.3.G.  .....6;.  
000030	5F 4B 87 62  8E 31 44 3E  3C 07 90 3A  91 12 71 82  _K.b.1D>  <..:..q.  
000040	6A 0C 8E 73  52 34 43 15  2D 70 15 65  8C 51 26 39  j..sR4C.  -p.e.Q&9  
000050	10 8C 8C 4F  23 6A 7A 94  2D 45 36 17  8B 6E 37 7E  ...O#jz.  -E6..n7~  
000060	7B 38 71 14  8A 5E 78 2C  10 22 1A 90  0D 8A 5E 28  {8q..^x,  ."....^(  
000070	32 09 49 93  55 43 7A 27  2F 87 2E 55  49 46 2C 44  2.I.UCz'  /..UIF,D  

000080	55 0F 4B 86  8E 0B 81 66  90 91 44 92  2A 0F 5B 53  U.K....f  ..D.*.[S  
000090	77 61 03 61  7D 6E 3E 6F  6C 7F 3D 4B  55 44 5F 7E  wa.a}n>o  l.=KUD_~  
0000A0	75 8B 7C 51  06 01 16 6E  2C 22 05 27  07 0F 2B 77  u.|Q...n  ,".'..+w  
0000B0	50 37 69 06  3F 34 1A 3A  63 51 90 26  6C 2F 33 63  P7i.?4.:  cQ.&l/3c  
```

**Comment by RobTillaart** [comment]
@brewmanz 
Merged develop branch into master and released as 0.3.0.  Please give SHEXA class a try.

If needed reopen the issue.
